api/src/main/java/io/grpc/NameResolver.java

```diff
@@ -50,7 +50,9 @@
  * <p>Implementations <strong>don't need to be thread-safe</strong>.  All methods are guaranteed to
  * be called sequentially.  Additionally, all methods that have side-effects, i.e.,
  * {@link #start(Listener2)}, {@link #shutdown} and {@link #refresh} are called from the same
- * {@link SynchronizationContext} as returned by {@link Args#getSynchronizationContext}.
+ * {@link SynchronizationContext} as returned by {@link Args#getSynchronizationContext}. <strong>Do
+ * not block</strong> within the synchronization context; blocking I/O and time-consuming tasks
+ * should be offloaded to a separate thread, generally {@link Args#getOffloadExecutor}.
```

Comment:
I added a comment in the original issue https://github.com/grpc/grpc-java/issues/8190#issuecomment-895620748 about using gRPC within the NameResolver which is more than just a blocking or long running operation inside the NameResolver. But this might. be sufficient.

Reply:
I responded on the issue. NR-using-gRPC happens, but I don't think that is the source of the current troubles. It doesn't matter if you are blocking using a gRPC API or any other, you really don't want to be doing that.